plugins/techdocs/src/reader/components/TechDocsReaderPage/TechDocsAuthProvider.tsx

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { ReactNode, useEffect, useState, useCallback } from 'react';
+import { ErrorPanel } from '@backstage/core-components';
+import { techdocsApiRef } from '@backstage/plugin-techdocs-react';
+import { useApi, useApp } from '@backstage/core-plugin-api';
+import useAsyncRetry from 'react-use/lib/useAsyncRetry';
+import { Button } from '@material-ui/core';
+
+type TechDocsRefreshCookieMessage = MessageEvent<{
+  action: string;
+  payload: {
+    expiresAt: string;
+  };
+}>;
+
+function useTechDocsCookie() {
```

Comment:
Arguably _everything_ could be done inside the hook.

```ts
function useCookieAuthRefresh(pluginId: string):
  | { state: 'loading' }
  | { state: 'ready' }
  | { state: 'error', error: Error }
```

That's all the caller needs to know about. We don't need to expose retries and stuff really.

And then the extension of that idea is - this hook could even be moved unchanged into `@backstage/plugin-auth-node` if we wanted to, for general reuse, when more plugins need to use it in the future. This assumes that everybody uses a well known cookie endpoint path, but it could be made into an option if absolutely needed. There's no real need to use the techdocs api ref as such. And it would be accompanied by a `<CookieAuthRefreshBarrier pluginId="techdocs">...` component too.

If you keep it together in one hook, it may also be easier to keep some small state at the top of the hook, and then have the different event mechanisms lined up clearly below. For example, if a channel event is received, just reset the expiresAt state to its new value, cancel your own current setTimeout that was waiting to start a refresh loop a few minutes before expiry, and start a new timeout and go back to sleep.

The cool thing also is, the timeout handle, the expiry time etc can all be useRef things. They don't need to cause repaints. The only thing that needs to cause a repaint is if the _output state_ changes (if we switch to an error state because nothing succeeded in refreshing a cookie in time).

Reply:
I really like this idea 🎉 
Do you mind if I do that in a followup pr?

Reply:
That's OK.

Reply:
Draft branch: https://github.com/backstage/backstage/compare/camilaibs/techdocs-auth-services-migration...camilaibs/create-generic-auth-cookie-provider?expand=1

Reply:
Here is the pull request: https://github.com/backstage/backstage/pull/23512